src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,140 @@
+/**
+ * @file batchnorm_impl.hpp
+ * @author Praveen Ch
+ * @author Manthan-R-Sheth
+ *
+ * Implementation of the Batch Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "batch_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+BatchNorm<InputDataType, OutputDataType>::BatchNorm(
+  const size_t size, const double eps) : size(size), eps(eps)
+{
+  weights.set_size(size + size, 1);
+}
+
+
+template<typename InputDataType, typename OutputDataType>
+void BatchNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+  deterministic = false;
+  gamma.fill(1.0);
+  beta.fill(0.0);
+  stats.reset();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BatchNorm<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  output.reshape(input.n_rows, input.n_cols);
+
+  // Mean and variance over the entire training set will be used to compute
+  // the forward pass when deterministic is set to true.
+  if (deterministic)
+  {
+    mean = stats.mean();
+    variance = stats.var(1);
+  }
+  else
+  {
+    mean = arma::mean(input, 1);
+    variance = arma::var(input, 1, 1);
+
+    for (size_t i = 0; i < output.n_cols; i++)
+    {
+      stats(input.col(i));
+    }
```

Comment:
I have a question about this.  In the paper, the authors simplify for the mini-batch setting:

```
In the batch setting where each training step is based on the entire training set, we would use the whole set to normalize activations. However, this is impractical when using stochastic optimization. Therefore, we make the second simplification: since we use mini-batches in stochastic gradient training, each mini-batch produces estimates of the mean and variance of each activation.
```

How does this apply to what you have implemented here?  To me, it looks like `stats` collects statistics on all points that pass through `Forward()` before `Reset()` is called.  (I believe `Reset()` is actually only called when the network is created or serialized, so maybe we should be resetting the running statistics more often?)

It would be possible to reset the statistics every call to `Forward()` because `Forward()` may be called with mini-batches, but, it might also be useful to find a way to allow the user to specify how often the statistics should be reset.  Let me know what you think---or if I misunderstood the functioning of the code here. :)

Reply:
`stats` object is used to get the mean and variance only during the testing time i.e. `deterministic = true`. So in the case of mini-batches at training, mean and variance are calculated from the samples in the mini-batch only, however they are stored in `stats` so that during testing, entire training statistics can be used.
Correct me if I misunderstood your question.

Reply:
Ah, now it makes sense; I overlooked how `mean` and `variance` are set differently.  Thank you for the clarification!